### PR TITLE
Uploading and downloading directory tree rather than archive

### PR DESCRIPTION
**Pull request by prismofeverything**
This solves a few problems we've been having and simplifies some other things:

* The namespace is now independent of the bucket name, which is now particular to each input or output key instead of globally declared. 
* Directories are now stored directly rather than archiving them and uploading the archive. This simplifies the layout and makes it easy for end users to download the entire tree or browse with the console or gcsfuse.
* Log output was coming through in batches of 32, now it displays each line as it arrives.

### CODE REVIEW

File: src/sisyphus/log.clj [hunk 1]
```diff
@@ -18,23 +18,23 @@
 
 (defn log!
   [^Level level & x]
-  (.log *logger* level (str x)))
+  (.log *logger* level (str x "\n")))
```
Comment (1fish2) [inline src/sisyphus/log.clj:21]: What needs a newline on every log entry? The current log handlers don't seem to.
I'll try the direct Stackdriver API soon.

File: src/sisyphus/task.clj [hunk 2]
```diff
@@ -144,6 +146,7 @@
   [{:keys [storage kafka docker config state]} task]
   (try
     (let [root (get-in config [:local :root])
+          user (or (System/getenv "USER") "root")
```
Comment (1fish2) [inline src/sisyphus/task.clj:149]: Is that the USER in the container? Is that why you want to call the run "root"?

Why get the docker container to log in as the end user? Is that the user's local login or their Google login? Wouldn't that make for some file permissions clashes?

Can we run all the worker processes as some consistent "sisyphus worker" user?

**Review by 1fish2 — COMMENTED** [review]
It's great to have these features.

A couple questions, below.

File: src/sisyphus/log.clj (hunk 1 above)
Comment (prismofeverything) [reply src/sisyphus/log.clj:21]: For some reason if I don't put the newline on there it all just prints on a single line in my terminal, and this fixed it. More mysterious logging features.

File: src/sisyphus/task.clj (hunk 2 above)
Comment (prismofeverything) [reply src/sisyphus/task.clj:149]: These are all excellent questions. You will see I have commented out the line below that actually adds the user to the container as it did not ultimately work (no "sisyphus" user inside the container), so I left this in to figure it out tomorrow.

Comment (prismofeverything) [reply src/sisyphus/task.clj:149]: > Is that the USER in the container? Is that why you want to call the run "root"?

The problem I am trying to solve is that as the container runs as `root`, all output files end up owned by `root` in the host system mounted folder (!) I was trying to get the USER from the outside environment (the user running sisyphus) to be the user who ultimately owns the output files. That is the effect I want, how to achieve this is still a mystery.

Comment (1fish2) [reply src/sisyphus/task.clj:149]: Ah! I see. I thought this was trying to identify the end user name who submitted the workflow, which is a sensible prefix name for the workflow's output files unless we decide to use it for the bucket name.

**Pull request merged**